File: app/src/main/java/com/salino/sali/ui/screens/history/HistoryViewModel.kt
package com.salino.sali.ui.screens.history

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.salino.sali.data.model.ShoppingItem
import com.salino.sali.domain.repository.AuthRepository
import com.salino.sali.domain.repository.ShoppingRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.Locale
import javax.inject.Inject

data class DayGroup(
    val dateLabel: String,
    val items: List<ShoppingItem>
)

data class HistoryState(
    val dayGroups: List<DayGroup> = emptyList(),
    val expandedDays: Set<String> = emptySet(),
    val items: List<ShoppingItem> = emptyList(),
    val isLoading: Boolean = true
)

@HiltViewModel
class HistoryViewModel @Inject constructor(
    private val shoppingRepository: ShoppingRepository,
    private val authRepository: AuthRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(HistoryState())
    val uiState: StateFlow<HistoryState> = _uiState

    private val dateFormatter = SimpleDateFormat("dd/MM/yyyy", Locale.getDefault())

    init {
        observeBoughtItems()
    }

    @OptIn(ExperimentalCoroutinesApi::class)
    private fun observeBoughtItems() {
        viewModelScope.launch {
            authRepository.observeCurrentUser()
                .filterNotNull()
                .collectLatest { user ->
                    val householdId = user.activeHouseholdId ?: return@collectLatest
                    shoppingRepository.observeBoughtItems(householdId).collect { items ->
                        val groups = items
                            .sortedByDescending { it.updatedAt?.toDate()?.time ?: 0L }
                            .groupBy { item ->
                                val date = item.updatedAt?.toDate()
                                if (date != null) dateFormatter.format(date) else ""
                            }
                            .map { (dateLabel, groupItems) -> DayGroup(dateLabel, groupItems) }

                        val expandedDays = _uiState.value.expandedDays

                        _uiState.value = HistoryState(
                            dayGroups = groups,
                            expandedDays = expandedDays,
                            items = items,
                            isLoading = false
                        )
                    }
                }
        }
    }

    fun toggleDay(dateLabel: String) {
        val current = _uiState.value.expandedDays
        _uiState.value = _uiState.value.copy(
            expandedDays = if (current.contains(dateLabel)) current - dateLabel else current + dateLabel
        )
    }

    fun returnToList(itemId: String) {
        viewModelScope.launch {
            val user = authRepository.observeCurrentUser().filterNotNull().first()
            val householdId = user.activeHouseholdId ?: return@launch
            shoppingRepository.markAsActive(householdId, itemId)
        }
    }
}

File: ios/ViewModels/HistoryViewModel.swift
import Foundation
import Combine

struct DayGroup: Identifiable {
    var id: String { dateLabel }
    var dateLabel: String
    var items: [ShoppingItem]
}

struct HistoryState {
    var dayGroups: [DayGroup] = []
    var expandedDays: Set<String> = []
    var items: [ShoppingItem] = []
    var isLoading = true
}

@MainActor
final class HistoryViewModel: ObservableObject {
    @Published var state = HistoryState()
    private let shoppingRepository: ShoppingRepository
    private let authRepository: AuthRepository
    private var householdId = ""
    private var task: Task<Void, Never>?

    init(shoppingRepository: ShoppingRepository, authRepository: AuthRepository) {
        self.shoppingRepository = shoppingRepository
        self.authRepository = authRepository
        task = Task { await observeBoughtItems() }
    }

    deinit { task?.cancel() }

    func toggleDay(_ dateLabel: String) {
        if state.expandedDays.contains(dateLabel) {
            state.expandedDays.remove(dateLabel)
        } else {
            state.expandedDays.insert(dateLabel)
        }
    }

    func returnToList(itemId: String) {
        Task { try? await shoppingRepository.markAsActive(householdId: householdId, itemId: itemId) }
    }

    private func observeBoughtItems() async {
        guard let user = await firstCurrentUser(from: authRepository), let activeHouseholdId = user.activeHouseholdId else {
            state.isLoading = false
            return
        }
        householdId = activeHouseholdId
        let formatter = DateFormatter()
        formatter.dateFormat = "dd/MM/yyyy"
        for await items in shoppingRepository.observeBoughtItems(householdId: activeHouseholdId) {
            let sorted = items.sorted { ($0.updatedAt ?? .distantPast) > ($1.updatedAt ?? .distantPast) }
            let grouped = Dictionary(grouping: sorted) { item in
                item.updatedAt.map { formatter.string(from: $0) } ?? ""
            }
            state.dayGroups = grouped
                .map { DayGroup(dateLabel: $0.key, items: $0.value) }
                .sorted { $0.dateLabel > $1.dateLabel }
            state.items = sorted
            state.isLoading = false
        }
    }
}
